library/alloc/src/vec/mod.rs

```diff
@@ -95,59 +94,149 @@ mod drain;
 #[cfg(not(no_global_oom_handling))]
 mod cow;
 
-#[cfg(not(no_global_oom_handling))]
 pub(crate) use self::into_iter::AsIntoIter;
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use self::into_iter::IntoIter;
 
 mod into_iter;
 
-#[cfg(not(no_global_oom_handling))]
 use self::is_zero::IsZero;
 
 mod is_zero;
 
-#[cfg(not(no_global_oom_handling))]
 mod source_iter_marker;
 
 mod partial_eq;
 
-#[cfg(not(no_global_oom_handling))]
 use self::spec_from_elem::SpecFromElem;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_from_elem;
 
-#[cfg(not(no_global_oom_handling))]
 use self::set_len_on_drop::SetLenOnDrop;
 
-#[cfg(not(no_global_oom_handling))]
 mod set_len_on_drop;
 
-#[cfg(not(no_global_oom_handling))]
 use self::in_place_drop::InPlaceDrop;
 
-#[cfg(not(no_global_oom_handling))]
 mod in_place_drop;
 
-#[cfg(not(no_global_oom_handling))]
 use self::spec_from_iter_nested::SpecFromIterNested;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_from_iter_nested;
 
-#[cfg(not(no_global_oom_handling))]
 use self::spec_from_iter::SpecFromIter;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_from_iter;
 
-#[cfg(not(no_global_oom_handling))]
 use self::spec_extend::SpecExtend;
 
-#[cfg(not(no_global_oom_handling))]
 mod spec_extend;
 
+trait Operations<T, A: Allocator> {
+    type Error;
+    fn shrink_to_fit(vec: &mut Vec<T, A>) -> Result<(), Self::Error>;
+    fn reserve(vec: &mut Vec<T, A>, additional: usize) -> Result<(), Self::Error>;
+    fn reserve_for_push(vec: &mut Vec<T, A>, len: usize) -> Result<(), Self::Error>;
+    fn with_capacity_in(capacity: usize, alloc: A) -> Result<Vec<T, A>, Self::Error>;
+    fn capacity_overflow<TReturn>() -> Result<TReturn, Self::Error>;
+    fn extend_with<E: ExtendWith<T>>(
+        vec: &mut Vec<T, A>,
+        n: usize,
+        value: E,
+    ) -> Result<(), Self::Error>;
+    fn with_capacity_zeroed_in(capacity: usize, alloc: A) -> Result<RawVec<T, A>, Self::Error>;
```

Comment:
```suggestion
    type Error;
    fn alloc_error(layout: Layout) -> Self::Error;
```
I have a feeling this is all you need! When creates the error, the other calls `handle_alloc_error`.

Then you can make private methods on the containers (rather than this trait) than take the `Operations` policy parameter, and do the right thing. Those private methods only call other such private methods, to ensure the policy is consistent. And then public methods wrap to hide the policy variable.

I think wit this strategy you can deduplicate *far* more code.

Reply:
Interesting... I'll try this now and let you know how it goes...

Reply:
That definitely simplifies things - it looks much better now!